servicetalk-data-protobuf-jersey/src/main/java/io/servicetalk/data/protobuf/jersey/ProtobufSerializerMessageBodyReaderWriter.java

```diff
@@ -0,0 +1,240 @@
+                           final MediaType mediaType, final MultivaluedMap<String, String> httpHeaders,
+                           final InputStream entityStream) throws WebApplicationException {
+        final ProtobufSerializerFactory serializerFactory = getSerializerFactory(mediaType);
+        final ExecutionContext<?> executionContext = ctxRefProvider.get().get().executionContext();
+        final BufferAllocator allocator = executionContext.bufferAllocator();
+        final int contentLength = requestCtxProvider.get().getLength();
+
+        if (Single.class.isAssignableFrom(type)) {
+            return handleEntityStream(entityStream, allocator,
+                    (p, a) -> deserialize(p, serializerFactory.serializerDeserializer(getSourceClass(genericType)),
+                            contentLength, a),
+                    (is, a) -> new SingleSource<>(deserialize(toBufferPublisher(is, a),
+                            serializerFactory.serializerDeserializer(
+                                    getSourceClass(genericType)), contentLength, a)));
+        } else if (Publisher.class.isAssignableFrom(type)) {
+            return handleEntityStream(entityStream, allocator,
+                    (p, a) -> serializerFactory.streamingSerializerDeserializer(
+                            getSourceClass(genericType)).deserialize(p, a),
+                    (is, a) -> new PublisherSource<>(serializerFactory.streamingSerializerDeserializer(
+                            getSourceClass(genericType)).deserialize(toBufferPublisher(is, a), a)));
+        }
+
+        return handleEntityStream(entityStream, allocator,
+                (p, a) -> deserializeObject(p, serializerFactory.serializerDeserializer(castClass(type)),
+                        contentLength, a),
+                (is, a) -> deserializeObject(toBufferPublisher(is, a),
+                        serializerFactory.serializerDeserializer(castClass(type)), contentLength, a));
+    }
+
+    @Override
+    public boolean isWriteable(final Class<?> type, final Type genericType, final Annotation[] annotations,
+                               final MediaType mediaType) {
+        return !isSse(requestCtxProvider.get()) && isSupportedMediaType(getRawWriteType(genericType), mediaType);
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @Override
+    public void writeTo(final Object o, final Class<?> type, final Type genericType,
+                        final Annotation[] annotations,
+                        final MediaType mediaType, final MultivaluedMap<String, Object> httpHeaders,
+                        final OutputStream entityStream) throws WebApplicationException {
+        final BufferAllocator allocator = ctxRefProvider.get().get().executionContext().bufferAllocator();
+        final Publisher<Buffer> bufferPublisher;
+        if (o instanceof Single) {
+            final Class<? extends MessageLite> clazz = genericType instanceof Class ? (Class) genericType :
```

Comment:
Why can it be assumed that the result will extend `MessageLite`? Is the `isWritable` sufficient? Perhaps worth an `assert class.isAssignableFrom(MessageLite.class) : "non-protobuf class class";`?

Reply:
the object maybe of type `javax.ws.rs.core.Response` which allows the user to customize headers and response code.

Reply:
`MessageLite` is the super class for all protobuf generated objects. The generics on the jersey serializer API doesn't easily allow for being more specific the types but if we encounter a class cast error a 5xx style error will be returned.

Reply:
OK